### PR TITLE
a simple way to gather profiling data

### PR DESCRIPTION
**Pull request by konnov**
The expression rewriter is keeping track of how many constants and expressions it has introduced in the SMT context. By connecting it to the source location, we produce a nice heat map. Though this profiling data is imprecise and definitely does not indicate the real complexity of the SMT problem, it actually highlights the weak spots of our encoding as well as hard parts of a TLA+ specification.

### CODE REVIEW

File: script/heatmap.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+# the regular expression for the source location as produced by SANY
+REGION_RE = re.compile(r"(.*):(\d+):(\d+)-(\d+):(\d+)")
+
+def region_from_csv(row):
+    m = REGION_RE.match(row[4])
+    if m:
+        entry = ProfileEntry(int(row[0]), int(row[1]), int(row[2]), int(row[3]))
+        return CodeRegion(m.group(1),
+                      (int(m.group(2)), int(m.group(3))), 
+                      (int(m.group(4)), int(m.group(5))),
```
Comment (shonfeder) [inline script/heatmap.py:74]: No essential, but I think you could improve readability here by using named groups in the regex and a `gruopdict` here: https://docs.python.org/3/library/re.html#re.Match.groupdict

File: script/heatmap.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    mw = 0
+    for r in regions:
+        mw = max(mw, r.profEntry.weight)
+
+    return mw
+
+
+def write_heatmap(heatfile, code_filename, regions):
+    """Write code lines with while coloring them by code regions"""
```
Comment (shonfeder) [inline script/heatmap.py:113]: I think there's a word missing here?

File: script/heatmap.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
```
Comment (shonfeder) [inline script/heatmap.py:114]: Seems to be an extra line?
```suggestion
```

File: script/heatmap.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    def split_line_by_regions(regions_ahead, line, line_no):
+        line_len = len(line)
+        # take the regions that cover the current line
+        window = itertools.takewhile(lambda r: r.fromPos[0] <= line_no, regions_ahead)
+        line_regions = sorted(list(window),
+                # first, sort by the starting column; then by the nr. of lines
+                key=lambda r: \
+                        (r.rangeByLine(line_no, line_len)[0], r.toPos[0] - r.fromPos[0]))
+
+        #print("Line: {}, length: {}".format(line_no, line_len))
```
Comment (shonfeder) [inline script/heatmap.py:129]: did you mean to leave these in for debugging purposes? (also on line 143 and 146)

File: script/heatmap.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+<html>
+  <head>
+    <style>
+      .code {
+        font-size: 12pt; font-family: "Courier New", Courier, monospace;
+      }
+    </style>
+  </head>
+  <body>
+    <h2>What is that?</h2>
```
Comment (shonfeder) [inline script/heatmap.py:238]:
```suggestion
    <h2>What is this?</h2>
```

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateRewriterImpl.scala [hunk 6]
```diff
@@ -394,7 +398,13 @@ class SymbStateRewriterImpl(private var _solverContext: SolverContext,
         state.setRex(eg._1)
 
       case None =>
+        // Get the SMT metrics before translating the expression.
+        // Note that we are not doing that in the recursive function,
+        // as the new expressions there will not have source information.
+        val smtWatermark = solverContext.metrics()
         val nextState = doRecursive(0, state)
+        profilerListener.foreach{ _.onRewrite(state.ex, solverContext.metrics().delta(smtWatermark)) }
+        // listener.recordMetrics(delta)
```
Comment (shonfeder) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateRewriterImpl.scala:407]: rm?

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/MetricProfilerListener.scala [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  * This listener registers the SMT metrics that are created when an expression is being translated in SMT.
+  * These metrics are collected only for those expressions that have source information, which can be
+  * displayed to the user.
+  *
+  * @param sourceStore the storage of source locations
+  * @param changeListener the tracer of expression updates
+  */
+class MetricProfilerListener(sourceStore: SourceStore, changeListener: ChangeListener, outFile: File)
+    extends SymbStateRewriterListener with LazyLogging {
+  private var _metricsPerId: Map[UID, SolverContextMetrics] = Map()
```
Comment (shonfeder) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/MetricProfilerListener.scala:21]: what is this one underscored?

**Review by shonfeder — APPROVED** [review]
LGMT

**Comment by lemmy** [comment]
FYI: The Toolbox's profiler frontend - that overlays a heatmap on top of the spec - expects simple plaintext input.  Let me know if you want to discuss the details to produce compatible output that can be read by the frontend.

File: script/heatmap.py (hunk 2 above)
Comment (konnov) [inline script/heatmap.py:113]:
```suggestion
    """Write code lines while coloring them by code regions"""
```

Comment (konnov) [reply script/heatmap.py:113]: Or one too many :-)

File: script/heatmap.py (hunk 1 above)
Comment (konnov) [reply script/heatmap.py:74]: Yeah. I think this script has to be reviewed at some point and simplified. I just wanted to try the idea. Dealing with code regions happened to be hard.

File: script/heatmap.py (hunk 4 above)
Comment (konnov) [reply script/heatmap.py:129]: Yes. These were quite useful. If there is a problem, we will need something similar.

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateRewriterImpl.scala (hunk 6 above)
Comment (konnov) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateRewriterImpl.scala:407]:
```suggestion
```

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/MetricProfilerListener.scala (hunk 7 above)
Comment (konnov) [reply tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/MetricProfilerListener.scala:21]: Sometimes, the underscored versions help to distinguish between the private attribute `x`, its getter (def `x`) and setter `def x_(newValue: T)...`. If you like to see more on scala getters and setter: https://stackoverflow.com/questions/26012039/scala-getters-setters-best-practice

However, it does not seem to be the case here. So it was more like a reflex to add underscore :-)

**Comment by konnov** [comment]
> FYI: The Toolbox's profiler frontend - that overlays a heatmap on top of the spec - expects simple plaintext input. Let me know if you want to discuss the details to produce compatible output that can be read by the frontend.

Yes! That sounds interesting. Is the file format documented? Btw, as far as I remember, TLC does profiling per action, whereas we do record metrics per expression (as all operators are inlined). Does it make a difference for the Toolbox?

**Comment by lemmy** [comment]
TLC profiles expressions (runtime & space) and actions (generated/unique states) in the behavior spec and the invariants, state-, and action-constraints.  As always, the code is the spec.

Below is the profiling for https://github.com/tlaplus/Examples/blob/master/specifications/ewd998/EWD998.tla with a few comments to understand the (human-readable) format.
```tla
<Init line 34, col 1 to line 34, col 4 of module EWD998>: 64:64   \* 64 states
  line 36, col 6 to line 36, col 34 of module EWD998: 1
  line 37, col 6 to line 37, col 31 of module EWD998: 8
  line 39, col 6 to line 39, col 34 of module EWD998: 64          \* Evaluated 64 times
  line 40, col 6 to line 40, col 34 of module EWD998: 64
  line 41, col 6 to line 41, col 52 of module EWD998: 64:4711     \* Evaluated 64 times, space of 4711 (space is made up here for illustrative purposes)
<InitiateProbe line 43, col 1 to line 43, col 13 of module EWD998>: 1857:1420024  \* unique/generated states
  line 45, col 6 to line 45, col 18 of module EWD998: 2140740
  |line 45, col 6 to line 45, col 14 of module EWD998: 1384582
  line 47, col 9 to line 47, col 29 of module EWD998: 1231052
  |line 47, col 9 to line 47, col 19 of module EWD998: 756268
  line 48, col 9 to line 48, col 26 of module EWD998: 1191764
  |line 48, col 9 to line 48, col 16 of module EWD998: 756268
  line 49, col 9 to line 49, col 32 of module EWD998: 509744
  |line 49, col 9 to line 49, col 28 of module EWD998: 756268
  line 50, col 6 to line 50, col 55 of module EWD998: 1420024
  line 51, col 6 to line 51, col 45 of module EWD998: 1420024
  line 53, col 6 to line 53, col 43 of module EWD998: 1420024
<PassToken line 55, col 1 to line 55, col 12 of module EWD998>: 51080:333672
  line 57, col 6 to line 57, col 16 of module EWD998: 3102836
  |line 57, col 8 to line 57, col 16 of module EWD998: 2769164
  line 58, col 6 to line 58, col 18 of module EWD998: 1887774
  |line 58, col 6 to line 58, col 14 of module EWD998: 1554102
  |line 58, col 18 to line 58, col 18 of module EWD998: 1554102
  line 59, col 6 to line 61, col 80 of module EWD998: 333672
  line 62, col 6 to line 62, col 45 of module EWD998: 333672
  line 64, col 6 to line 64, col 43 of module EWD998: 333672
<SendMsg line 71, col 1 to line 71, col 10 of module EWD998>: 384467:3814706
  line 73, col 6 to line 73, col 14 of module EWD998: 1907353
  |line 73, col 6 to line 73, col 11 of module EWD998: 4153746
  |line 73, col 13 to line 73, col 13 of module EWD998: 4153746
  line 75, col 6 to line 75, col 45 of module EWD998: 1907353
  line 77, col 29 to line 77, col 68 of module EWD998: 3814706
  line 77, col 15 to line 77, col 25 of module EWD998: 1907353
  line 80, col 6 to line 80, col 39 of module EWD998: 3814706
<RecvMsg line 82, col 1 to line 82, col 10 of module EWD998>: 227576:2674524
  line 83, col 6 to line 83, col 19 of module EWD998: 2674524
  |line 83, col 6 to line 83, col 15 of module EWD998: 4153746
  line 84, col 6 to line 84, col 45 of module EWD998: 2674524
  line 86, col 6 to line 86, col 45 of module EWD998: 2674524
  line 88, col 6 to line 88, col 45 of module EWD998: 2674524
  line 90, col 6 to line 90, col 44 of module EWD998: 2674524
  line 91, col 6 to line 91, col 24 of module EWD998: 2674524
<Deactivate line 93, col 1 to line 93, col 13 of module EWD998>: 719538:1907353
  line 94, col 6 to line 94, col 14 of module EWD998: 1907353
  |line 94, col 6 to line 94, col 11 of module EWD998: 4153746
  |line 94, col 13 to line 94, col 13 of module EWD998: 4153746
  line 95, col 6 to line 95, col 43 of module EWD998: 1907353
  line 96, col 6 to line 96, col 49 of module EWD998: 1907353
<Inv line 149, col 1 to line 149, col 3 of module EWD998>
  line 150, col 3 to line 162, col 34 of module EWD998: 2821262
  |line 150, col 11 to line 150, col 45 of module EWD998: 2821262
  |line 153, col 6 to line 162, col 34 of module EWD998: 2821262
  ||line 153, col 14 to line 156, col 74 of module EWD998: 2821262
  |||line 153, col 17 to line 153, col 56 of module EWD998: 2821262
  ||||line 153, col 46 to line 153, col 56 of module EWD998: 3438829
  ||||line 153, col 26 to line 153, col 43 of module EWD998: 2821262
  |||line 154, col 17 to line 156, col 74 of module EWD998: 1028209
  ||||line 154, col 20 to line 154, col 34 of module EWD998: 1028209
  ||||line 155, col 22 to line 155, col 32 of module EWD998: 147358
  ||||line 156, col 22 to line 156, col 74 of module EWD998: 880851
  ||line 158, col 14 to line 158, col 64 of module EWD998: 2493264
  ||line 160, col 14 to line 160, col 55 of module EWD998: 728458
  |||line 160, col 38 to line 160, col 55 of module EWD998: 795554
  |||line 160, col 23 to line 160, col 34 of module EWD998: 728458
  ||line 162, col 14 to line 162, col 34 of module EWD998: 150134
<TypeOK line 26, col 1 to line 26, col 6 of module EWD998>
  line 27, col 3 to line 31, col 20 of module EWD998: 2821262
<constr_160461242609934000 line 10, col 1 to line 10, col 25 of module MC>: 8713663:10150343
  line 11, col 1 to line 11, col 15 of module MC: 10150343
  |line 113, col 3 to line 114, col 17 of module EWD998: 1
0150343
  ||line 113, col 6 to line 113, col 56 of module EWD998: 10150343
  |||line 113, col 23 to line 113, col 56 of module EWD998: 28793617
  ||||line 113, col 23 to line 113, col 37 of module EWD998: 28793617
  ||||line 113, col 42 to line 113, col 56 of module EWD998: 27825949
  |||line 113, col 15 to line 113, col 19 of module EWD998: 10150343
  ||line 114, col 6 to line 114, col 17 of module EWD998: 8713663
End of statistics.
```

**Comment by konnov** [comment]
Cool! I will open a separate issue for that.

**Pull request merged**